### PR TITLE
fix: use Fifo queues and set message group id

### PR DESCRIPTION
**Pull request by ReubenRobertsHEE**
There are a lot of line changes, but mostly its just swopping-out the default `QueueMessagingTemplate.convertAndSend()` to rather use the new `FifoMessagingService.sendMessageToFifoQueue()`. This figures out the message group id and sends the message.

The sketchiest bit is figuring out the message group id for the data requests. As it happens, the requests are all for 'primary' objects, which can be handled quite simply, except for requesting a placement's primary PlacementSpecialty, but it is rather brittle.

Likewise, the '`is it a Record, or extended from a Record, or one of these specifically coded classes`' logic for calculating the message group id in the FifoMessagingService is quite tightly coupled to the existing class design, which will probably change quite soon.

TIS21-6040

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -106,3 +106,7 @@ tasks.test {
   finalizedBy(tasks.jacocoTestReport)
   useJUnitPlatform()
 }
+
+tasks.checkstyleMain {
```
Comment (ReubenRobertsHEE) [inline build.gradle.kts:110]: To review: this is not a great solution, but not sure if having a different version of the plugin for this component is better.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestService.java [hunk 2]
```diff
@@ -56,22 +59,52 @@ public DataRequestService(QueueMessagingTemplate messagingTemplate,
...
-  public String sendRequest(String tableName, Map<String, String> whereMap)
+  public String sendRequest(String schema, String tableName, Map<String, String> whereMap)
       throws JsonProcessingException {
     String messageBody = makeJson(tableName, whereMap);
 
-    log.info("Sending SQS message with body: [{}]", messageBody);
-    messagingTemplate.convertAndSend(queueUrl, messageBody);
+    String tisId = whereMap.values().toArray()[0].toString();
+    //note: ordering cannot be guaranteed, but only a single value map is ever provided except for
+    //the exception PlacementSpecialty handled below.
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestService.java:77]: This also feels very ad-hoc.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestService.java [hunk 3]
```diff
@@ -56,22 +59,52 @@ public DataRequestService(QueueMessagingTemplate messagingTemplate,
...
+    Map<String, Object> headers = new HashMap<>();
+    //All data requests are for primary (parent) table records, so they can be left as-is,
+    //except for a PlacementSpecialty request which fortunately has the parent placement's id.
+    if (tableName.equalsIgnoreCase("PlacementSpecialty")) {
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestService.java:81]: ditto.

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.4.2-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.4-bin.zip
```
Comment (ReubenRobertsHEE) [inline gradle/wrapper/gradle-wrapper.properties:3]: Upgrade needed since the checkstyle plugin choked on the new FifoMessagingService code.

**Comment by ReubenRobertsHEE** [comment]
Sorry, I see the gradle update has unearthed a good number of code smells (in existing code). These could be resolved...

**Review by Judge40 — APPROVED** [review]
If we weren't going to be taking another look at sync in general I might be wanted to come up with some more elegant solutions for some of the fiddly bits. But assuming that work is imminent this LGTM

**Pull request merged**